File: llm-python/security/audit/service.py
import logging
from collections.abc import Callable


class AuditService:
    def __init__(self, writer: Callable[[object], None] | None = None):
        self._writer = writer

    def write_event(self, event) -> None:
        if self._writer is not None:
            self._writer(event)
            return
        logging.info(
            "audit service=%s action=%s result=%s details=%s",
            event.service,
            event.action,
            event.result,
            event.details,
        )

File: portal-go/internal/security/audit/service.go
package audit

import (
	"context"
	"log"

	"portal-go/internal/security"
)

type LogWriter interface {
	WriteSecurityLog(ctx context.Context, event security.AuditEvent) error
}

type Service struct {
	writer LogWriter
}

func New(writer LogWriter) *Service {
	return &Service{writer: writer}
}

func (s *Service) WriteEvent(ctx context.Context, event security.AuditEvent) error {
	if s.writer != nil {
		return s.writer.WriteSecurityLog(ctx, event)
	}
	log.Printf("audit service=%s action=%s result=%s details=%s", event.Service, event.Action, event.Result, event.Details)
	return nil
}
